src/execution_plan/execution_plan.c

```diff
@@ -885,39 +873,44 @@ ExecutionPlan *NewExecutionPlan(ResultSet *result_set) {
 		}
 
 		// Prepare column names for the ResultSet.
-		if(result_set) ResultSet_SetColumns(result_set, AST_BuildColumnNames(last_clause));
+		plan->returned_column_names = AST_BuildColumnNames(last_clause);
 
-		OpBase *results_op = NewResultsOp(plan, result_set);
+		OpBase *results_op = NewResultsOp(plan);
 		_ExecutionPlan_UpdateRoot(plan, results_op);
 	} else if(last_clause_type == CYPHER_AST_CALL) {
 		assert(plan->root->type == OPType_PROC_CALL);
 		OpProcCall *last_op = (OpProcCall *)plan->root;
 		// Prepare column names for the ResultSet.
-		if(result_set) {
-			const char **columns;
-			array_clone(columns, last_op->output);
-			ResultSet_SetColumns(result_set, columns);
-		}
+		array_clone(plan->returned_column_names, last_op->output);
 
-		OpBase *results_op = NewResultsOp(plan, result_set);
+		OpBase *results_op = NewResultsOp(plan);
 		_ExecutionPlan_UpdateRoot(plan, results_op);
 	}
 
-	// Optimize the operations in the ExecutionPlan.
-	optimizePlan(plan);
-
 	// Disregard self.
 	plan->segment_count = segment_count - 1;
 	plan->segments = segments;
-	QueryCtx_SetLastWriter(_ExecutionPlan_FindLastWriter(plan->root));
+
 	return plan;
 }
 
+void ExecutionPlan_PreparePlan(ExecutionPlan *plan) {
+	optimizePlan(plan);
+	QueryCtx_SetLastWriter(_ExecutionPlan_FindLastWriter(plan->root));
+	if(plan->returned_column_names) ResultSet_SetColumns(QueryCtx_GetResultSet(),
```

Comment:
Let's have `ResultSet_SetColumns` invoke `QueryCtx_GetResultSet` rather than expecting it as an argument.